### PR TITLE
Add reranker component

### PR DESCRIPTION
**Pull request by bhavani105**
# Pull Request      

## What does this PR do?

This adds a Reranker component
gRPC and REST endpoints for reranking
Two  Rerankers
- ColBERT Reranker
- SeqClassificationReranker

Notes:
- Replace `(issue)` **above** ↑↑↑ with the issue this PR closes to automatically link the two.
This must be done when the PR is created.
- Add multiple `Closes #(issue)` as needed.
- If this PR is work towards but does not close an issue, simply tag the issue without mentioning `Closes`.

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

```(description)```     

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [ ] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [ ] Added [description](#description) of changes proposed.
- [ ] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [ ] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

File: primeqa/components/reranker/colbert_reranker.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        default=None,
+        metadata={
+            "name": "Checkpoint",
+            "description": "Path to checkpoint",
+        },
+    )
+    '''
+
+    model: str = field(
+        default="drdecr", #"cross-encoder/ms-marco-TinyBERT-L-2",
```
Comment (seirasto) [inline primeqa/components/reranker/colbert_reranker.py:41]: remove this comment?

File: primeqa/components/reranker/colbert_reranker.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    Args:
+        model (str, optional): Model to load.
+        collection (str, optional): collection to load. Defaults to collection in index configuration.
+        max_num_documents (int, optional): Maximum number of reranked document to return. Defaults to 5.
+
+    Important:
+    1. Each field has metadata property which can carry additional information for other downstream usages.
+    2. Two special keys (api_support and exclude_from_hash) are defined in "metadata" property.
+        a. api_support (bool, optional): If set to True, that parameter is exposed via service layer. Defaults to False.
```
Comment (seirasto) [inline primeqa/components/reranker/colbert_reranker.py:23]:
```suggestion
        a. api_support (bool, optional): If set to True, that parameter is exposed via the service layer. Defaults to False.
```

File: primeqa/components/reranker/colbert_reranker.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+class ColBERTReranker(BaseReranker):
+    """_summary_
+
+    Args:
+        model (str, optional): Model to load.
+        collection (str, optional): collection to load. Defaults to collection in index configuration.
+        max_num_documents (int, optional): Maximum number of reranked document to return. Defaults to 5.
+
+    Important:
+    1. Each field has metadata property which can carry additional information for other downstream usages.
```
Comment (seirasto) [inline primeqa/components/reranker/colbert_reranker.py:21]:
```suggestion
    1. Each field has the metadata property which can carry additional information for other downstream usages.
```

File: primeqa/components/reranker/colbert_reranker.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+                        "text": "A man is eating food.",
+                        "document_id": "0",
+                        "title": "food"
+                    },
+                    "score": 1.4
+                }
+
+        Returns:
+            List[List[Dict]] For each query a list of reranked documents in the same
+            structure as the input documents with the score replace with the reranker score.
```
Comment (seirasto) [inline primeqa/components/reranker/colbert_reranker.py:127]: Suggested replacement for the two lines: 

List[List[Dict]] A list of reranked documents in the same structure as the input documents with the score replaced with the reranker score for each query.

File: primeqa/services/exceptions.py [hunk 5]
```diff
@@ -36,6 +36,9 @@ class ErrorMessages(str, Enum):
     FAILED_TO_LOCATE_INDEX_INFORMATION = (
         "E6003: Index information for index with id {} doesn't exist."
     )
+    
+    # RETRANKER
+    INVALID_RETRANKER = "E5001: Invalid reranker: {}. Please select one of the following pre-defined rerankers: {}"
 
     # INITIALIZATION
     FAILED_TO_INITIALIZE = "E9001: Failed to initalize {}. Please contact us."
```
Comment (seirasto) [inline primeqa/services/exceptions.py:44]:
```suggestion
    FAILED_TO_INITIALIZE = "E9001: Failed to initialize {}. Please contact us."
```

File: primeqa/services/grpc_server/server.py [hunk 6]
```diff
@@ -71,6 +73,11 @@ def run(self) -> None:
             retriever_pb2_grpc.add_RetrievingServiceServicer_to_server(
                 RetrieverService(config=self._config), server
             )
+            
+            # Add retriever service
```
Comment (seirasto) [inline primeqa/services/grpc_server/server.py:77]:
```suggestion
            # Add reranker service
```

File: tests/primeqa/services/rest_server/test_rerankeddocuments.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
```
Comment (jdpsen) [inline tests/primeqa/services/rest_server/test_rerankeddocuments.py:1]: we should rename it to test_reranked_documents.py

File: primeqa/components/base.py [hunk 8]
```diff
@@ -155,3 +155,44 @@ def get_engine_type(self) -> str:
...
+            queries (List[str]): search queries
+            texts (List[List[Dict]]): For each query, a list of documents to rerank
+                where each document is a dictionary with the following structure:
+                {
+                    "document": {
+                        "text": "A man is eating food.",
+                        "document_id": "0",
+                        "title": "food"
+                    },
+                    "score": 1.4
```
Comment (jdpsen) [inline primeqa/components/base.py:191]: do we necessarily need the score parameter to be present in the doc dict ? Are we using it in our re-ranker ?

**Review by jdpsen — CHANGES_REQUESTED** [review]
Added minor comments in the files. 
Also couple of other points:
1. We should add a Re-ranker section in the ReadMe https://github.com/primeqa/primeqa/blob/add_reranker_component/primeqa/components/README.md listing the two re-rankers we are using and 
2. a notebook to show how to use the re-ranker component. 

The code changes look good to me.

File: primeqa/components/reranker/colbert_reranker.py (hunk 4 above)
Comment (franzmpub) [reply primeqa/components/reranker/colbert_reranker.py]: Updated as suggested

File: primeqa/components/reranker/seq_classification_reranker.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+    Args:
+        model (str, optional): Model to load. 
+        collection (str, optional): collection to load. Defaults to collection in index configuration.
+        max_num_documents (int, optional): Maximum number of reranked document to return. Defaults to 5.
+
+    Important:
+    1. Each field has metadata property which can carry additional information for other downstream usages.
+    2. Two special keys (api_support and exclude_from_hash) are defined in "metadata" property.
+        a. api_support (bool, optional): If set to True, that parameter is exposed via service layer. Defaults to False.
```
Comment (seirasto) [inline primeqa/components/reranker/seq_classification_reranker.py:26]:
```suggestion
        a. api_support (bool, optional): If set to True, that parameter is exposed via the service layer. Defaults to False.
```

File: primeqa/ir/dense/colbert_top/colbert/searcher.py [hunk 10]
```diff
@@ -16,13 +16,26 @@
...
+TextDocuments = Union[List[str]]
 
 class Searcher:
-    def __init__(self, index, checkpoint=None, collection=None, config=None):
+    def __init__(self, index, checkpoint=None, collection=None, config=None, rescore_only=False):
         print_memory_stats()
-        
+
         initial_config = ColBERTConfig.from_existing(Run().config, config)
+        self.rescore_only = rescore_only
```
Comment (seirasto) [inline primeqa/ir/dense/colbert_top/colbert/searcher.py:26]: Can you add some comments in this file?

**Review by seirasto — COMMENTED** [review]
I've left some comments (mostly suggestions for comments/typos). Can you also add details about this to the README?

**Review by franzmpub — COMMENTED** [review]
Proposed one trivial change, otherwise LGTM

**Review by jdpsen — APPROVED** [review]
Please correct the minor typos. The notebooks can be done later after the demo.
Approving to get this merged.

**Review by seirasto — APPROVED** [review]
LGTM. Please see my comment regarding args that can possibly be in base class and address if needed when you think is appropriate.

**Review by franzmpub — APPROVED** [review]
LGTM

**Pull request merged**